PLMS.BLL/ServicesImplementation/TaskService.cs

```diff
@@ -1,206 +1,338 @@
+                t => t.Goal.Category,
+                t => t.TaskComments);
+
+            AddTaskDto taskDto = new()
+            {
+                Title = task.Title,
+                Description = task.Description ?? string.Empty,
+                GoalId = task.Goal?.Id ?? 0,
+                TaskComments = task.TaskComments?.Select(
                     t => new TaskCommentDto
                     {
                         Comment = t.Comment,
                         CreatedAt = t.CreatedAt
-                    }),
+                    })?.ToArray() ?? Array.Empty<TaskCommentDto>(),
                 CreatedAt = task.CreatedAt,
                 DueDate = task.DueDate,
                 PriorityId = task.Priority.Id,
                 StatusId = task.Status.Id,
-                CategoryId = task.Goal.Category.Id
             };
 
             return taskDto;
         }
+
+        private async Task<Goal> MapToGoalAsync(AddTaskDto taskDto, IRepository<Category> categoryRepository, IRepository<Priority> priorityRepository, IRepository<Status> statusRepository,
+            Priority defaultPriority, Status defaultStatus, string userId)
+        {
+            taskDto.Goal.UserId = userId;
+            Goal goal = _mapper.Map<Goal>(taskDto.Goal);
+            var category = await categoryRepository.GetByIdAsync(taskDto.Goal.CategoryId);
+            if (category == null)
+            {
+                taskDto.Goal.Category.UserId = userId;
+                category = _mapper.Map<Category>(taskDto.Goal.Category);
+            }
+            goal.Category = category;
+            goal.Priority = await priorityRepository.GetByIdAsync(taskDto.Goal.PriorityId) ?? defaultPriority;
+            goal.Status = await statusRepository.GetByIdAsync(taskDto.Goal.StatusId) ?? defaultStatus;
+
+            return goal;
+        }
+
+        private async Task<IEnumerable<Task>> GetFilteredTasksAsync(TaskItemFilter filters)
+        {
+            var filterExpression = ApplyFilters(filters);
+            var orderFunction = ApplyOrder(filters);
+
+            // Apply filtation and sorting on the DB level.
+            var taskRepository = _unitOfWork.GetRepository<Task>();
+            var query = taskRepository.GetAll()
+                                       .Include(t => t.Goal)
+                                       .ThenInclude(g => g.Category)
+                                       .Include(t => t.Goal)
+                                       .ThenInclude(g => g.GoalComments)
+                                       .Include(t => t.Priority)
+                                       .Include(t => t.Status)
+                                       .Include(t => t.TaskComments)
+                                       .Where(filterExpression);
+
+            var orderedQuery = orderFunction(query);
+
+            // Pagination
+            var paginatedQuery = orderedQuery.Skip((filters.PageNumber - 1) * filters.ItemsPerPageCount)
```

Comment:
Very elegant implementation of the method.
Once it's possible to check in the runtime, verify that skip/take are part of the SQL select query (that they are executed on LINQ to SQL, not on LINQ to Object)